File: app/src/main/java/com/hinterlong/kevin/swishticker/ui/modules/HistoryFragment.java
package com.hinterlong.kevin.swishticker.ui.modules;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.hinterlong.kevin.swishticker.QueryEngine;
import com.hinterlong.kevin.swishticker.R;
import com.hinterlong.kevin.swishticker.ui.adapters.GameItem;

import java.util.ArrayList;
import java.util.List;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.RecyclerView;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;
import eu.davidea.flexibleadapter.FlexibleAdapter;

/**
 * A fragment representing a list of Items.
 */
public class HistoryFragment extends Fragment {
    private final FlexibleAdapter<GameItem> adapter = new FlexibleAdapter<>(null);
    @BindView(R.id.game_list) RecyclerView recyclerView;
    private Unbinder unbinder;

    public HistoryFragment() {

    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_game_list, container, false);
        unbinder = ButterKnife.bind(this, view);
        recyclerView.setAdapter(adapter);

        return view;
    }

    @Override
    public void onResume() {
        super.onResume();
        adapter.updateDataSet(getGamesList());
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        adapter.onSaveInstanceState(outState);
        super.onSaveInstanceState(outState);
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        adapter.onRestoreInstanceState(savedInstanceState);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        unbinder.unbind();
    }

    public List<GameItem> getGamesList() {
        List<GameItem> list = new ArrayList<>();
        for (int gameId : QueryEngine.getInstance().getGameIds()) {
            GameItem gameItem = new GameItem(QueryEngine.getInstance().getGame(gameId), gameId);
            list.add(gameItem);
        }
        return list;
    }
}

File: app/src/main/java/com/hinterlong/kevin/swishticker/ui/modules/HistoryFragment.kt
package com.hinterlong.kevin.swishticker.ui.modules

import android.os.Bundle
import android.os.Handler
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.Transformations
import com.hinterlong.kevin.swishticker.R
import com.hinterlong.kevin.swishticker.service.AppDatabase
import com.hinterlong.kevin.swishticker.service.gameScores
import com.hinterlong.kevin.swishticker.ui.adapters.GameItem
import com.hinterlong.kevin.swishticker.utilities.Prefs
import eu.davidea.flexibleadapter.FlexibleAdapter
import kotlinx.android.synthetic.main.fragment_game_list.*
import kotlinx.android.synthetic.main.game_item.view.*


/**
 * A fragment representing a list of Items.
 */
class HistoryFragment : Fragment() {
    private val adapter = FlexibleAdapter<GameItem>(null)

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_game_list, container, false)

        return view
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        gamesList.adapter = adapter

        Transformations.map(AppDatabase.getInstance(view.context)
            .gameDao()
            .getGamesAndActions(), ::gameScores)
            .observe(viewLifecycleOwner, Observer {
                adapter.notifyItemRangeRemoved(0, adapter.itemCount)
                adapter.updateDataSet(it.map { GameItem(it.game, it.score, viewLifecycleOwner) })
                adapter.notifyItemRangeChanged(0, it.size)

                if (it.size == 1 && Prefs.isFirstDemoDeleteGame) {
                    demoDeleteAction()
                    Prefs.isFirstDemoDeleteGame = false
                }
            })
    }

    private fun demoDeleteAction() {
        val handler = Handler()
        handler.postDelayed({
            val swipeLayout = gamesList.findViewHolderForLayoutPosition(0)?.itemView?.swipeLayout
            swipeLayout?.animateSwipeRight()
            Toast.makeText(context, "Swipe right to delete games", Toast.LENGTH_SHORT).show()
            handler.postDelayed({
                swipeLayout?.animateReset()
            }, 800)

        }, 500)
    }


    override fun onSaveInstanceState(outState: Bundle) {
        adapter.onSaveInstanceState(outState)
        super.onSaveInstanceState(outState)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        adapter.onRestoreInstanceState(savedInstanceState)
    }
}
